### PR TITLE
obs-scripting: Switch Python support to use the Py3 stable ABI

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
Changes the Python module of obs-scripting to use the Python 3 stable ABI and also allows OBS to load several different recent Python 3 versions (any possible version between 3.6 and 3.10 will be loaded, if available).

Also changes the macOS UI to just require the base path a `Python.framework` is placed in (instead of requiring users to dive into the Framework to find the library path).

**Note:** This PR implements the changes introduced by https://github.com/obsproject/obs-studio/pull/6696 to pull the swig binaries compiled on CI with stable Abu support.

### Motivation and Context
In theory these changes should allow users to run Python scripts in OBS with a wider range of available Python versions installed on the host system, independent from the version used to compile OBS.

Caveats: While the obspython module uses the stable ABI, obs-scripting itself cannot do the same, as OBS uses C-API methods not available in the stable ABI. As the use of these methods is overseen by OBS development, additional care can be put into their implementation, ensuring that only methods available/unchanged in respective Python versions are used.

The stable ABI itself does not contain many C-API methods that _should_ be contained, for the simple reason of not being as actively maintained/updated as Python itself. Hence why we chose a "better is the enemy of good" approach.

### How Has This Been Tested?
Compiled and ran OBS with Python scripts on:
* Windows (Python 3.6 and 3.9)
* Linux (Python 3.10)
* macOS (Python 3.9 and Python 3.10)

### Types of changes
- Tweak (non-breaking change to improve existing functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Review by RytoEX — APPROVED** [review]
Seems fine to me at a glance. Will need to merge #6696 first.

**Comment by WizardCM** [comment]
During my own testing, the CI artifact of this build only worked with Python 3.10 on Windows. Both 3.6 and 3.9 fail to load.

**Comment by RytoEX** [comment]
# Python 3.6

Using Python 3.6.2 installed via pyenv-win doesn't seem to load anything at all. Scripts report "no properties available" and nothing appears in the Script Log. The OBS log reports:
```
22:52:47.032: [Python] Could not load library: C:/pyenv-win/versions/3.6.2/
22:54:01.254: [Python] Could not load library: C:/pyenv-win/versions/3.6.2/Lib/
22:54:18.071: [Python] Python failure in get_defaults:168:
```


# Python 3.9

Using Python 3.9 installed with the default Python installer into Program Files, OBS loads Python, but fails to execute my test script, claiming that "obs_frontend_preview_program_mode_active" does not exist. This reminds me of #6539 where Lua scripts became unable to access Frontend APi functions that were not _explicitly_ made visible to scripts, where previously we relied on some automagic to make them available. See #6553 for how that was addressed at the time.

My test script also says that OBS_FRONTEND_EVENT_STREAMING_STARTING is undefined, which is similar to the linked Issue.

# Python 3.10

With Python 3.10 installed with the default Python installer, I get the same results described in the Python 3.9 section above.

**Comment by PatTheMav** [comment]
> During my own testing, the CI artifact of this build only worked with Python 3.10 on Windows. Both 3.6 and 3.9 fail to load.

Yeah that's an oversight, I fixed this locally already. I can confirm the issues reported by @RytoEX on Windows (macOS works fine).

**Comment by PatTheMav** [comment]
Just pushed an update that should address both issues.

File: deps/obs-scripting/obspython/CMakeLists.txt [hunk 1]
```diff
@@ -21,8 +21,10 @@ endif()
 include(UseSWIG)
 
 set_source_files_properties(
-  obspython.i PROPERTIES USE_TARGET_INCLUDE_DIRECTORIES TRUE SWIG_FLAGS
-                                                             "-builtin;-py3")
+  obspython.i
+  PROPERTIES USE_TARGET_INCLUDE_DIRECTORIES TRUE
+             SWIG_FLAGS
+             "$<IF:$<BOOL:${OS_LINUX}>,-builtin;-py3,-py3-stable-abi>")
```
Comment (RytoEX) [inline deps/obs-scripting/obspython/CMakeLists.txt:27]: The `-py3` flag is deprecated and no longer supported.

**Review by RytoEX — CHANGES_REQUESTED** [review]
It looks like the use of `-py3` is emitting a deprecation warning due to SWIG dropping the flag.

https://github.com/swig/swig/commit/a343b7e254567a64761bc1be7dc55b7b7424ec52

**Review by RytoEX — APPROVED** [review]
Tests on Windows look good to me.

**Comment by WizardCM** [comment]
To better clarify the above from Ryan:

I was running into two separate issues. In my opinion, **neither is a blocker for this PR** to be merged but at the very least both should be independently fixed later.

### 1. Python 3.8.0 and 3.8.1 load successfully, but fail to import the obspython.py/pyd

In this situation, the following is logged on startup or on first attempt of loading 3.8.0 or 3.8.1.

```
21:13:25.988: [Python] Python failure in obs_scripting_load_python:1724:
21:13:25.988: [Python] Error importing obspython.py', unloading obs-python
21:13:25.988: [Python] Failed to load python plugin
```

At this point in the code, the Python interpreter is fully loaded and can run Python code, but when it attempts to load our `obspython.py` it spits out the following into stderr, which is not exposed in any way and provides no additional insight. The error itself comes from the Python interpreter.

```
ImportError: DLL load failed while importing obspython: The specified module could not be found.
```

* For users, the fix is to update to 3.8.2 or above.

* For us, we should (at the very least) display some kind of error in the UI that our Python implementation failed to load, and we **should not save** the selected path.

### 2. If Python loads successfully but obspython doesn't, attempting to load a broken _or_ functional Python+obspython a second time causes OBS to crash

I have a feeling this is caused by two separate internal issues, one is related to `python_loaded_at_all` (which is false when it should be true) and `mutexes_loaded` being true when `tick_mutex` and `timer_mutex` are null, as issue (1) skips half the loading steps. There may be one more related to how other "python_is_loaded" variables function because of this.

Either way, without bypassing the `mutexes_loaded` conditional, I get the following crash with the CI build. https://obsproject.com/logs/6Z5oUwFtdYUkYgrL

Steps:

1. Load Python 3.8.0 or 3.8.1
2. Load *any* version fo Python, including 3.6, 3.8 or 3.10
3. Crash

**Comment by PatTheMav** [comment]
From what I could tell the issue in Python 3.8.0 and Python 3.8.1 is related to Py 3.8 changing the way DLLs are loaded (or more precisely, which paths Python supplies as search paths for the library loader):

Before Py 3.8, Python used the `PATH` as possible locations for finding DLLs (and as such DLLs linked to from precompiled python modules), which allowed DLL hijacking. For security reasons this was removed, now the loading application needs to supply DLL search paths using Windows APIs.

I haven't figured out yet what changed in that regard with Python 3.8.2 (which doesn't seem to exhibit the same issue).

**Pull request merged**